chore: update pre-commit hooks

wyardley opened the pull request:
Note: the current `ruff` pre-commit checks check `.`, however, there are multiple existing errors. Maybe it would make sense to switch `ruff` to run on changed files only unless `--all-files` is passed? That would be gentler, especially if we're not requiring all these checks to pass (that said, #958 has some misc fixes that resolve _some_ of the outstanding flagged and non-ignored issues).
 
- Update pre-commit hooks
- Switch away from deprecated flag to `ruff`
- Resolve two minor pre-commit hook findings

Note: I still see the following finding from `mypy`:
```
semantic_release/cli/commands/publish.py: note: In function "publish":
semantic_release/cli/commands/publish.py:66:9: error: "HvcsBase" has no
attribute "upload_dists"  [attr-defined]
            hvcs_client.upload_dists(tag=tag, dist_glob=pattern)
            ^~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 49 source files)
```
Maybe an inheritance issue with it being in `remote_hvcs_base.py` vs `_base.py`?

There is also some commented out code that ruff is flagging - I can remove in another PR if you'd like.

wyardley commented on semantic_release/cli/config.py:
```diff
@@ -558,7 +558,7 @@ def from_raw_config(
             parts = [*env_var_def.split("=", 1), ""]
             # removes any odd spacing around =, and extracts name=value
-            name, env_val = [part.strip() for part in parts[:2]]
+            name, env_val = (part.strip() for part in parts[:2])
```
This was added by `pyupgrade`

codejedi365 replied:
makes sense

wyardley commented on .pre-commit-config.yaml:
```diff
@@ -48,27 +48,27 @@ repos:
           - "--exit-non-zero-on-fix"
           - "--statistics"
-          - "--output-format=text"
+          - "--output-format=full"
```
updated elsewhere in b41ad31793e5ee02359e8949cc0dd5d69240b622 I think.

codejedi365 replied:
yes that is correct

wyardley commented on .pre-commit-config.yaml:
```diff
@@ -31,44 +31,38 @@ repos:
+      - id: ruff-format
+        name: ruff (format)
+        args: ["."]
+        pass_filenames: false
```
I think removing these two lines for both lint and fmt should switch it back to the more traditional behavior of only running on modified files, which IMO is the right behavior for a pre-commit hook (esp. since the checks run on everything in CI anyway)

codejedi365 replied:
concur. only lets keep it the same version that is in `pyproject.toml` for now, and then also see if there is a dependabot configuration for this file.  That is how we keep ruff updated in `pyproject.toml`. There is a PR for `ruff` currently from dependabot but i haven't merged it yet because honestly `ruff` is releasing so frequently it's annoying.  I'm slow rolling the updates here for this dependency.  Just for awareness, I have it locked to the specific patch version because if there is a bug fix that re-enables a rule it would arbitrarily cause the CI to break upon their release.  I'm not okay with that so I'll just slow roll the merges as its only a dev dependency.

wyardley replied:
I don't think the hook version corresponds directly to the ruff version, though they may still be linked.

Not 100% sure on dependabot, but FWIW, Renovate does have experimental support (which works quite well) for keeping track of pre-commit hook versions.

I saw the comment, but I actually think (could be wrong) that having the old version of the pre-commit pinned here might be part of why people are seeing some fighting sometimes (as maybe the hook uses an internal version of `ruff`) tied to its own version?

I do see errors with the current ruff version if I run `ruff check` locally as well, though (note: using Python 3.12 vs. 3.8, but the toml config does have target version set to 3.8) - do you not see that?

wyardley replied:
per https://github.com/python-semantic-release/python-semantic-release/pull/958#issuecomment-2170391124, adjusted to operate only on changed files. will double-check the behavior.

codejedi365 commented on action.sh:
Looking at the chmod value, please do not add execute to this. It is not intended to be run in someone's environment and I don't like laying down executable shell scripts on someones box via git.  Making it executable is not needed here because the Dockerfile calls `bash action.sh` which means it doesnt require execution permissions but still runs as intended.

wyardley replied:
Ok - I was basing this on the existing pre-commit hooks existing. So I think if you don't want this behavior, the repo should not have those lines in the pre-commit config either. I removed them and made the script non-executable.

codejedi365 replied:
Good catch, yeah I haven't migrated over to use `pre-commit` since becoming maintainer of this project so I was unaware where that was coming from.

wyardley commented on pyproject.toml:
```diff
@@ -282,7 +282,6 @@ ignore = [
 ]
 
 external = ["V"]
-ignore-init-module-imports = true
```
this resolves this deprecation warning:
```
warning: The `ignore-init-module-imports` option is deprecated and will be removed in a future release. Ruff's handling of imports in `__init__.py` files has been improved (in preview) and unused imports will always be flagged.
```

wyardley commented:
However, the hook version doesn't seem to affect the version of `ruff` run. _if_ I use `--all-files`, I do still see a couple errors with `ruff-lint`, as well as one `mypy` error that doesn't seem to show up in CI. Either way, I think this will solve at least some of the problems with running it locally, but let me know if you want to pin back the pre-commit hook to match the ruff version.

codejedi365 reviewed (approved):
I bumped the project ruff version so that matches what is in the pre-commit definition. And then rebased this branch so that they are the same.

Also I did review the `ruff-pre-commit` releases to find that each version corresponds to which ruff it uses under the hood.

codejedi365 commented:
> However, the hook version doesn't seem to affect the version of `ruff` run. _if_ I use `--all-files`, I do still see a couple errors with `ruff-lint`, as well as one `mypy` error that doesn't seem to show up in CI. Either way, I think this will solve at least some of the problems with running it locally, but let me know if you want to pin back the pre-commit hook to match the ruff version.

I really believe this is because the pre-commit is not reading our config in the pyproject.toml file because if you see the lint job in the ci we have to specify the configuration file to ruff. Why it doesn't do this by default is besides me but it changes what is flagged in the CI

wyardley commented:
Also, just read that the lint should come before format if lint is run with `—fix`, so maybe I should switch the order?

wyardley commented:
Also, I can try adding that argument here and see if that fixes the issue.

wyardley commented:
> Also, I can try adding that argument here and see if that fixes the issue.

So, just FWIW, I checked and adding `--config=pyproject.toml` to the args for pre-commit config doesn't seem to change the behavior (also, I noticed that ruff does seem to pick up settings from `pyproject.toml` even when it's not called with that argument).  If it matters, I'm also not seeing codes for the specific findings (for example, `ERA001`) ignored in the config file.

Playing around with it locally, I _think_ the actual issue is that having the `--diff` flag set seems to suppress some of the findings. If you try `ruff check --fix` vs `ruff check --fix --diff` 

```shell
$ ruff check --fix
[...]
tests/unit/semantic_release/hvcs/test_gitlab.py:
  612:5 SIM117 Use a single `with` statement with multiple contexts instead of nested `with` statements

tests/util.py:
  15:34 N812 Lowercase `config` imported as non-lowercase `cliConfigModule`

Found 26 errors.
$ ruff check --diff
$ echo $?
0
```

So even though there are findings, `ruff check --diff` seems to be exiting 0 in any case (note: I think it's not printing anything else because there are no _fixable_ findings, even though there are findings).

That said, we can leave this as it is here, and I can try ignoring and / or chipping away some of those other findings in another PR.

wyardley commented:
I'm trying to fix the mypy pre-commit issue with `--all-files` too, but may have to figure that out separately. It basically has to do with the way files are provided / passed 

Just setting `semantic-release/` as a final `args` value doesn't seem to do it properly, probably because it expects filenames to be passed in last as a positional argument, but with `pre-commit run --all-files`, it's throwing at least one error

```
% mypy semantic_release                                                         
Success: no issues found in 46 source files
% python3 -m mypy --config pyproject.toml --ignore-missing-imports semantic_release
Success: no issues found in 46 source files
% pre-commit run --all-files
[...]
semantic_release/cli/commands/publish.py:66:9: error: "HvcsBase" has no
attribute "upload_dists"  [attr-defined]
            hvcs_client.upload_dists(tag=tag, dist_glob=pattern)
            ^~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 46 source files)
```

Another thing I can do separately, but I _think_ with the versions of `mypy` and `ruff` we have, we may not need to set the config file anymore for either?

wyardley commented:
Threw up a draft first pass in #961 at the rest of the ruff failures and removing `--diff` from the CI job

codejedi365 commented:
> % python3 -m mypy --config pyproject.toml --ignore-missing-imports semantic_release
> Success: no issues found in 46 source files
> % pre-commit run --all-files
> [...]
> semantic_release/cli/commands/publish.py:66:9: error: "HvcsBase" has no
> attribute "upload_dists"  [attr-defined]
>             hvcs_client.upload_dists(tag=tag, dist_glob=pattern)
>             ^~~~~~~~~~~~~~~~~~~~~~~~
> Found 1 error in 1 file (checked 46 source files)
> ```

Let's just throw a `# type: ignore` on this because this will be refactored soon anyway. 

> Another thing I can do separately, but I _think_ with the versions of `mypy` and `ruff` we have, we may not need to set the config file anymore for either?

Probably but not sure. The less command line arguments the better.

wyardley commented:
> Probably but not sure. The less command line arguments the better.

Ok, see how you think this looks now - removed what I think should be unnecessary args, also added that ignore.

codejedi365 commented:
Thanks 🎉
